content/shmem_query_interoperability.tex

```diff
@@ -0,0 +1,50 @@
+\apisummary{
+    Determines whether an interoperability feature is supported by the \openshmem
```

Comment:
Is there a general prototype of this routine? Has this been tested? Sorry, I haven't been focusing on this proposal before - in general, I'm bit confused on the implementation of such a routine.

Lets consider a OpenSHMEM implementation, with the following scenario:
1. It can guarantee SHMEM_PROGRESS_MPI for an MPI implemetation MPI:1
2. It can't guarantee SHMEM_PROGRESS_MPI for an MPI implementation MPI:2
Also, both MPI:1 and MPI:2 are third party implementations and the OpenSHMEM implementation have no control over it. We can fill any known MPI and OpenSHMEM implementations in this scenario.

AFAIU, reading environment variables from the MPI implementations is incorrect and there are no standard MPI routines to determine the version info like ```shmem_info_get_name```.  So, how would the OpenSHMEM implementation determine the linked MPI implementation? Please clarify.

Reply:
CASE-1: If MPI is provided by a third party implementation, I do not think the OpenSHMEM implementation can guarantee progress for MPI. In such a case, the return value of `SHMEM_PROGRESS_MPI` should be `0` in general. However, this feature may be supported for such a scenario if the implementations agree on a special synchronization protocol (**vendor-specific**). For example:
 - The MPI and OpenSHMEM implementations share the same process manager.
 - They handshake at initialization through the process manager and the MPI implementation exposes a progress hook.
 - If the OpenSHMEM impl already uses a software thread, it can trigger the MPI progress hook to ensure MPI progress.
 - All the above conditions can be confirmed once both MPI and OpenSHMEM are initialized, thus the query API can return the corresponding value.

CASE-2: The support of `SHMEM_PROGRESS_MPI` can be easier if MPI and OpenSHMEM are provided by the same runtime system. The implementation can ensure that the user is using its MPI and OpenSHMEM once the program called `MPI_Init` and `shmem_init` . For example:
- The runtime can set internal flags `is_my_mpi_initialized=1` at  `MPI_Init` and set `is_my_shmem_initialized=1` at `shmem_init`.
- When the user queries  `SHMEM_PROGRESS_MPI`, return `is_my_mpi_initialized && is_my_shmem_initialized ? 1:0`.

I have tested the API in OSHMPI which is similar to CASE-2.

Reply:
Discussion at today's thread WG meeting:
It is more suitable to expose the progress support of MPI through an MPI API rather than an OpenSHMEM API. Because we do not have a useful case for the query API now, I will postpone the query API part and only go with the first section of the proposal.